juju/model.py

```diff
@@ -0,0 +1,166 @@
+        self.relations = RelationMap(data['relations'])
+
+    def relation(self, relation_name):
+        """Return the first Relation object for the named relation, or None.
+
+        This is a convenience method for the case where only a single related
+        app is supported. If there are more than one remote apps on the relation,
+        it will raise a TooManyRelatedApps model error. If no there are no remote
+        apps on the relation, it will return None. Otherwise, it is equivalent
+        to self.relations[relation_name][0].
+        """
+        num_related = len(self.relations[relation_name])
+        if num_related > 1:
+            raise TooManyRelatedApps(relation_name, num_related, 1)
+        elif num_related == 0:
+            return None
+        else:
+            return self.relations[relation_name][0]
+
+
+class Application:
+    def __init__(self, name, units):
+        # TODO Do we want to move to UUIDs? Is that even possible with the number of
+        # places that rely on unit names being of the form {app_name}/{unit_number}?
+        # Note that CMRs don't expose the actual unit name, but still follow the form,
+        # using someething like remote-56a0f163eb4e4f2e88d50983dca7be02/0.
+        self.name = name
+        self.units = units
+
+    def __hash__(self):
+        # An Application instance from one relation will not be identical to one from another relation,
+        # since they may have different views of the units of that app. However, they should be considered
+        # equivalent, since they are views of the same entity. Generally, this should only matter for using
+        # model.app as an index into a given relation's relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Application):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
+        return self.name == other.name
+
+
+class Unit:
+    def __init__(self, name):
+        # TODO Do we want to move to UUIDs? Is that even possible with the number of
+        # places that rely on unit names being of the form {app_name}/{unit_number}?
+        # Note that CMRs don't expose the actual unit name, but still follow the form,
+        # using someething like remote-56a0f163eb4e4f2e88d50983dca7be02/0.
+        self.name = name
+
+    def __hash__(self):
+        # A Unit instance from one relation will not be identical to one from another relation.
+        # However, they should be considered equivalent, since they are views of the same entity.
+        # Generally, this should only matter for using model.unit as an index into a given relation's
+        # relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Unit):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
```

Comment:
I think should just be `return False`. That's the typical behavior of Python on equality across different types.